src/lib/comms/sol-connman-impl-connman.c

```diff
@@ -0,0 +1,906 @@
+{
+    int id;
+    struct sol_connman_service *service;
+
+    if (_ctx.connman) {
+        sol_bus_client_free(_ctx.connman);
+        _ctx.connman = NULL;
+    }
+
+    _ctx.monitor_callback.cb = NULL;
+    _ctx.monitor_callback.data = NULL;
+
+    _ctx.state_slot.slot =
+        sd_bus_slot_unref(_ctx.state_slot.slot);
+    _ctx.manager_slot.slot =
+        sd_bus_slot_unref(_ctx.manager_slot.slot);
+    _ctx.service_slot.slot =
+        sd_bus_slot_unref(_ctx.service_slot.slot);
+
+    SOL_VECTOR_FOREACH_IDX(&_ctx.service_vector, service, id) {
+        _free_connman_service(service);
+    }
+
+    sol_vector_clear(&_ctx.service_vector);
+
+    _ctx.connman_state = SOL_CONNMAN_STATE_UNKNOWN;
+}
+
+static int
+_match_properties_changed(sd_bus_message *m, void *userdata,
+    sd_bus_error *ret_error)
+{
+    const char *interface;
+    int r;
+
+    SOL_NULL_CHECK(_ctx.properties_changed, -EINVAL);
+
+    interface = sd_bus_message_get_interface(m);
+    SOL_NULL_CHECK(interface, -EINVAL);
+
+    if (strncmp(interface, "net.connman.", 12) != 0)
+        return -EINVAL;
+
+    if (sd_bus_message_is_signal(m, "net.connman.Manager",
+            "ServicesChanged")) {
+        if (_ctx.service_slot.slot)
+            return -EINVAL;
+
+        services_list_changed(m);
+    } else if (sd_bus_message_is_signal(m, "net.connman.Manager",
+            "PropertyChanged")) {
+        char *str = NULL;
+
+        if (_ctx.manager_slot.slot)
+            return -EINVAL;
+
+        r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &str);
+        SOL_INT_CHECK(r, < 0, r);
+        if (strcmp(str, "State") == 0) {
+            get_manager_properties(m);
+        } else {
+            r = sd_bus_message_skip(m, "v");
+            SOL_INT_CHECK(r, < 0, r);
+        }
+    }
+
+    return 0;
+}
+
+static int
+dbus_connection_add_monitor(
+    void (*cb)(void *data, const struct sol_connman_service *service),
+    const void *data)
+{
+    int r;
+    char matchstr[256];
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+
+    if (_ctx.properties_changed)
+        return 0;
+
+    _ctx.monitor_callback.cb = cb;
+    _ctx.monitor_callback.data = data;
+
+    r = snprintf(matchstr, sizeof(matchstr),
+            "type='signal',"
+            "interface='net.connman.Manager'");
+    SOL_INT_CHECK(r, < 0, -EINVAL);
+
+    r = sd_bus_add_match(bus, &_ctx.properties_changed, matchstr,
+            _match_properties_changed, NULL);
+    SOL_INT_CHECK(r, < 0, -EINVAL);
+
+    return 0;
+}
+
+static void
+dbus_connection_del_monitor(
+    void (*cb)(void *data, const struct sol_connman_service *service),
+    const void *data)
+{
+
+}
+
+SOL_API int
+sol_connman_add_service_monitor(
+    void (*cb)(void *data, const struct sol_connman_service *service),
+    const void *data)
+{
+    int r;
+
+    r = sol_connman_init_lazy();
+    SOL_INT_CHECK_GOTO(r, < 0, fail_init);
+
+    r = dbus_connection_get_manager_properties();
```

Comment:
What about if the user calls `sol_connman_add_service_monitor()` twice, it will send all these messages and try to parse them again. And if the user doesn't add any monitor, he/she will never get any meaningful information.

What I propose, is that you add watches for `Manager.ServicesChanged`, `Manager.PropertiesChanged` and each service `Service.PropertiesChanged`, and `call 'GetServices()' (did I miss anything?) during the module initialization. So if the user later calls any method, even if there's no monitor added, the user will get a valid and meaningful information. What do you think?